bootstrap/src/main/java/io/airlift/bootstrap/LifeCycleManager.java

```diff
@@ -233,7 +235,27 @@ private void startInstance(Object obj)
     {
         log.debug("Starting %s", obj.getClass().getName());
         LifeCycleMethods methods = methodsMap.get(obj.getClass());
-        for (Method postConstruct : methods.methodsFor(PostConstruct.class)) {
+
+        Collection<Method> postConstructMethods = methods.methodsFor(PostConstruct.class);
+        boolean hasPostConstructMethods = !postConstructMethods.isEmpty();
+        boolean hasPreDestroyMethod = methodsMap.get(obj.getClass()).hasFor(PreDestroy.class);
+
+        if (!hasPostConstructMethods && !hasPreDestroyMethod) {
+            // no need to track in startedInstances or managedInstances
+            return;
+        }
+
+        // Guice can double provision instances (in particular with Optional binding).
+        // Protect against calling post-construct and pre-destroy methods more than once
+        if (startedInstances.putIfAbsent(identityHashCode(obj), obj) != null) {
+            return;
+        }
+
+        if (hasPreDestroyMethod) {
+            managedInstances.add(obj);
+        }
+
+        for (Method postConstruct : postConstructMethods) {
```

Comment:
```suggestion
       for (Method postConstruct : methods.methodsFor(PostConstruct.class)) {
```

Reply:
Why call `methods.methodsFor(PostConstruct.class)` twice? Are you sure the optimizer will hoist this?

Reply:
this doesn't matter, the number of such annotated objects is small, I prefer readable and concise code here